File: src/leetCodeProblems/Greedy/LongestConsecutiveSequence128.java
package leetCodeProblems.HashSearch;

/**
 * LeetCode - https://leetcode.com/problems/longest-consecutive-sequence/
 * TimeComplexity - O(n2)
 * SpaceComplexity - O(n)
 */

import java.util.*;

public class LongestConsecutiveSequence128 {

	public int longestConsecutive(int[] nums) {
        HashSet<Integer> hashSet = new HashSet<Integer>();
        int ans = 0;

        for(int i=0; i < nums.length; i++) {
            hashSet.add(nums[i]);
        }

        for(int i=0; i < nums.length; i++) {

            // This is important, to remove UNREQUIRED loops.
            if (!hashSet.contains(nums[i]-1)) {

                int j = 0;

                while(hashSet.contains(nums[i]+j)) {
                    j++;
                }

                if (j > ans) {
                    ans = j;
                }
            }
        }

        return ans;
    }

    public static void main(String[] args) {

        LongestConsecutiveSequence128 obj = new LongestConsecutiveSequence128();

        int[] nums = {100,4,200,1,3,2};

        System.out.println(obj.longestConsecutive(nums)); // Expected output 4
    }
}

File: src/leetCodeProblems/Greedy/LongestConsecutiveSequence128.go
package main

/*
- Leetcode - https://leetcode.com/problems/longest-consecutive-sequence
- Time - O(n)
- Space - O(n)
*/
func longestConsecutive(nums []int) int {
	m := make(map[int]struct{})

	for _, v := range nums {
		m[v] = struct{}{}
	}

	out := 0

	for v := range m {
		if _, ok := m[v-1]; ok {
			continue
		}

		len := 0

		for {
			if _, ok := m[v+len]; !ok {
				break
			}
			len++
		}

		if len > out {
			out = len
		}
	}
	return out
}
